onnx/defs/controlflow/defs.cc

```diff
@@ -272,6 +272,107 @@ void IfInferenceFunction(InferenceContext& ctx) {
   }
 }
 
+void LoopInferenceFunction(InferenceContext& ctx) {
+  auto num_inputs = ctx.getNumInputs();
+  auto num_loop_state_vars = num_inputs - 2; // skip 'M' and 'cond'
+
+  std::vector<const TypeProto*> subgraph_input_types;
+  subgraph_input_types.push_back(nullptr); // iteration number. not inferred.
+  subgraph_input_types.push_back(ctx.getInputType(1)); // 'cond' value.
+
+  // loop state values get propagated to outputs
+  for (size_t i = 2; i < num_inputs; ++i) {
+    auto input_type = ctx.getInputType(i);
+    subgraph_input_types.push_back(input_type);
+
+    propagateElemTypeFromInputToOutput(ctx, i, i - 2);
+
+    if (hasInputShape(ctx, i)) {
+      propagateShapeFromInputToOutput(ctx, i, i - 2);
+    }
```

Comment:
My reading of the op documentation is that the scan-outputs are required to have a shape that is invariant across the loop iterations … but I don't see a similar restriction imposed on the loop-state variables. (We do have such a restriction in the case of "scan"). I would drop the above 3 lines (for loop-state vars) unless the op. spec is changed.

Reply:
Likewise the type pushed into subgraph_input_types should be a copy of the type with shape information omitted, for the same reason, since this shape could vary over iterations.